samples/java-spring-choreography-saga-quickstart/src/main/java/user/registry/entities/UserEntity.java

```diff
@@ -0,0 +1,155 @@
+package user.registry.entities;
+
+
+import kalix.javasdk.StatusCode;
+import kalix.javasdk.annotations.*;
+import kalix.javasdk.eventsourcedentity.EventSourcedEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.bind.annotation.*;
+import user.registry.Done;
+
+import java.util.List;
+
+/**
+ * Entity representing a User.
+ * <p>
+ * A User has a name, a country and an email address.
+ * The email address must be unique across all existing users. This is achieved with a choreography saga which ensures that
+ * a user is only created if the email address is not already reserved.
+ * <p>
+ * This entity is protected from outside access. It can only be accessed from within this service (see the ACL annotation).
+ * External access is gated and should go through the ApplicationController.
+ */
+@Id("id")
+@TypeId("user")
+@RequestMapping("/users/{id}")
+@Acl(allow = @Acl.Matcher(service = "*"))
+public class UserEntity extends EventSourcedEntity<UserEntity.User, UserEntity.Event> {
+
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  // commands
+  public record Create(String name, String country, String email) {
+  }
+
+  public record ChangeEmail(String newEmail) {
+  }
+
+  /**
+   * It's recommended to seal the event interface.
+   * As such, Kalix can detect that there are event handlers defined for each event.
+   */
+  public sealed interface Event {
+  }
+
+  @TypeName("user-created")
+  public record UserWasCreated(String name, String country, String email) implements Event {
+  }
+
+  @TypeName("email-assigned")
+  public record EmailAssigned(String newEmail) implements Event {
+  }
+
+  @TypeName("email-unassigned")
+  public record EmailUnassigned(String oldEmail) implements Event {
+  }
+
+  public record User(String name, String country, String email) {
```

Comment:
```
Also, reinforce the fact that this domain is encapsulated by the Entity wrapper.
```
What is the added value? I understand that encapsulating the domain class itself is very important, but I don't see the advantage of wrapping the code in the Entity. I would say that in 90% of well-structured projects, you will find a  `domain` package with domain code.

```
But my idea is not to suggest any pattern or good practices.
```
My impression is that whatever coding pattern/practice we suggest in our samples, most likely will be followed/replicated.

```
The goal is just to let the user open one file and see what are all the pieces that make this entity.
```
That is only your perception. For some people, this might be overwhelming, too much code in one place, hard to grasp what is what, and hard to differentiate where is the responsibility bondary. For others, this might be just fine. I'm in the first group, but it doesn't matter. What matters is where the average Java developer is (in which group). My intuition is that it is still the first group because, for years, one class = one file was the Java standard.

Reply:
LGTM 👍

Reply:
> Two more cents from my side. We should gather more data points for some concerns, like packaging, code conventions etc. We could ask about such things Java architects who are reviewing Kalix.

Well, we can, but at some point we should just decide on one approach and be consistent. If we ask 10 people, we will get 10 different opinions. At the end, what is important is that we make something that newcomers can find themselves. Each developers will have their own taste and will encode their application they way they prefer.

> 
> Second, we discussed the idea of promoting onion architecture, I think that quickstarts are a good place to actually do this.

When we will have a page on the docs about this. Once we have it, we can add a link from the README to the architecture style page.